packages/core/src/conflict-resolution.ts

```diff
@@ -174,31 +167,45 @@ export async function fetchLog(
   dispatcher: Dispatcher,
   cid: CID,
   stateLog: HistoryLog,
-  log: CID[] = [],
-): Promise<CID[]> {
-  if (await stateLog.includes(cid)) {
+  log: LogEntry[] = [],
+): Promise<LogEntry[]> {
+  if (stateLog.includes(cid)) {
     // already processed
     return [];
   }
   const commit = await dispatcher.retrieveCommit(cid);
-  if (commit == null) {
-    throw new Error(`No commit found for CID ${cid.toString()}`);
-  }
+  if (!commit) throw new Error(`No commit found for CID ${cid.toString()}`);
 
   let payload = commit;
-  if (StreamUtils.isSignedCommit(commit)) {
+  const isSignedCommit = StreamUtils.isSignedCommit(commit);
+  if (isSignedCommit) {
     payload = await dispatcher.retrieveCommit(commit.link);
-    if (payload == null) {
-      throw new Error(`No commit found for CID ${commit.link.toString()}`);
-    }
+    if (!payload) throw new Error(`No commit found for CID ${commit.link.toString()}`);
   }
   const prevCid: CID = payload.prev;
   if (!prevCid) {
     // this is a fake log
     return [];
   }
-  log.push(cid); // Should be unshift [O(N)], but push [O(1)] + reverse [O(N)] seem better
-  if (await stateLog.includes(prevCid)) {
+  let nextEntry: LogEntry;
+  if (isSignedCommit) {
+    nextEntry = {
+      type: CommitType.SIGNED,
+      cid: cid,
+    };
+  } else {
+    // If not signed, then anchor commit
```

Comment:
Ehm, it could also be a genesis commit? Or is that handled by line 172?

Reply:
It is combination of 172 plus a caller side. State that is passed there is already constructed from the genesis.

Reply:
Working note: This is true only for Tiles. Caip10 has a completely different notion of a signed commit.